File: src/main/java/org/karel/karel/compiler/BeepersMap.java
package org.karel.karel.compiler;

import java.util.HashMap;
import java.util.Map;
import java.util.Objects;

public class BeepersMap {
    private final Map<Coordinates, Integer> beepers;

    public BeepersMap(Map<Coordinates, Integer> beepers) {
        this.beepers = beepers;
    }

    public BeepersMap(){
        this(new HashMap<>());
    }

    public int getBeepersCount(int x, int y){
        if (beepers.get(new Coordinates(x, y)) == null){
            return 0;
        } else {
            return beepers.get(new Coordinates(x, y));
        }
    }

    public int getBeepersCount(Coordinates c){
        return getBeepersCount(c.getX(), c.getY());
    }

    public Map<Coordinates, Integer> getBeepers() {
        return new HashMap<>(beepers);
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (!(o instanceof BeepersMap that)) return false;
        return Objects.equals(beepers, that.beepers);
    }

    @Override
    public int hashCode() {
        return Objects.hashCode(beepers);
    }
}

File: src/main/resources/static/js/BeepersMap.js
export class BeepersMap {
    constructor(beepers = new Map()) {
        this.beepers = beepers;
    }

    getBeepersCount(x, y) {
        return this.beepers.get(`${x},${y}`) || 0;
    }

    getBeepersCountByCoordinates(coordinates) {
        return this.getBeepersCount(coordinates.x, coordinates.y);
    }

    getBeepers() {
        return new Map(this.beepers);
    }

    equals(other) {
        if (!(other instanceof BeepersMap)) return false;
        return this.beepers.size === other.beepers.size && [...this.beepers.entries()].every(([key, value]) => other.beepers.get(key) === value);
    }

    hashCode() {
        return JSON.stringify([...this.beepers]);
    }
}
